### ISSUE TITLE
Optional setter return value

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxe#7507]
If the setter does not return anything, maybe the compiler should return the parameter that was passed automatically.

**Comment by nadako** [comment]
I support this. I would maybe go even further and deprecate returning the value from a setter, because it's a very bad practice to return anything other than the input value. It can be extra-confusing when using assignment chains.

**Comment by ncannasse** [comment]
@nadako I don't think we should go as far, and assignation has some kind of semantics.

For instance:

```
var i : Int;
var u8 : UInt8;
i = (u8 = 456);
trace(i); //  200 
```

**Comment by nadako** [comment]
@ncannasse I think that is unrelated, because it's a normal type change that can also happen with field setting, e.g.:

```
var s:{i:Int} = ...;
var u8:UInt8;
trace(s.i = (u8 = 456)); // 200
```

Haxe will first coerce the result value of `u8=456` to `Int` and then assign it to `s.i`. So, if `s.i` was a void-setter, we'd generate something like:
```
var tmp:Int = (u8 = 456); // 200 already here
s.set_i(tmp);
trace(tmp);
```

The point is, all type coercion is (or at least should be) done by normal unification before using the value for the field assignment/setter call and the field set operation shouldn't have any type-changing semantics by itself.

> PS unless i'm misunderstanding something here, of course :)

**Comment by ncannasse** [comment]
Well, anyway we don't want a warning here for people that have lot of existing code or want to keep haxe 3.x compatibility. So let's make it optional and update the documentation so that there's no return by default.

**Comment by Simn** [comment]
I think it's time to add `-w`.

**Comment by RblSb** [comment]
I think it is not the parameter should be returned, but the field itself?
```haxe
function set_id(value:Int) {
_id = value + 1;
return _id; //can sometimes be replaced by an argument in the optimization phase?
}
```
Also I have a question, should the setter return something? I mean, just generate getter if we are trying to get the value from setter.
https://try.haxe.org/#c4d97
Is it possible to execute code in "block 1" in the same way as code in "block 2", so that the results are identical? If this can be considered a correct behavior, how much does it affect compatibility?

**Comment by nadako** [comment]
> I think it is not the parameter should be returned, but the field itself?
> 
> ```hx
> function set_id(value:Int) {
> _id = value + 1;
> return _id; //can sometimes be replaced by an argument in the optimization phase?
> }
> ```

This is a perfect example of the code that shouldn't exist and could be avoided if Haxe didn't support arbitrary return values from the setter :)

Having said that though, maybe always emitting a warning would indeed be too annoying for the existing code, so just changing docs and pretending like it was never a feature works for me :) Also, it *might* have a legit use case for some specific DSL like query builders and whatnot.

**Comment by RblSb** [comment]
@nadako there may also be a value conversion in the setter, for example for externs, probably. Anyway, i am not against the return of the argument, because I do not come across such cases, just talking about their possibility.

**Comment by ncannasse** [comment]
I don't think we should return a call to the getter. And yes there's still
Dan example, which would be rewritten as the following under new syntax:

```haxe
function set_id(value:Int) {
value++;
_id = value;
}
```

But uhm I don't like it much either. We could still make the `return`
mandatory when there's an actual getter, but return the field if it's
`default` access.

**Comment by nadako** [comment]
Having to deal with setter returns in my as3->haxe converter a bit reminded me of this issue. Does anyone have any feelings with regard to it? Should we maybe just do it, e.g. for 4.1?

**Comment by Simn** [comment]
The problem I have with this is that a setter doesn't really know that it's a setter. We use a naming convention for this, but the design has always been that setters are just normal functions.

A setter doesn't even have to live on the same class as its field once inheritance gets involved. This means that adding a property to a child class could suddenly transform a setter in a base class, which is just weird.

I would prefer to not touch this and instead introduce a better property syntax altogether. I believe there's an open haxe-evolution about that.

**Comment by kLabz** [comment]
> I believe there's an open haxe-evolution about that.

Yep, https://github.com/HaxeFoundation/haxe-evolution/pull/63

**Comment by nadako** [comment]
> adding a property to a child class could suddenly transform a setter in a base class

Probably I'm misunderstanding how it works, but I thought it would be implemented not by transforming anything but when typing setter calls: if the setter call returns `Void` - use the passed value, otherwise use the result of setter call. That way we'd support both forms.

**Comment by Simn** [comment]
I'm not sure, `a = b = c` is `a = (b = c)`, but with this you would transform it to `{ b = c; a = c; }`. I think this could change semantics if the `b = c` assignment causes a coercion. But that probably escapes Haxe-defined semantics.

**Comment by nadako** [comment]
But if `b = c` is a Void-returning setter, it can't change semantics? (and if it's a value-returning setter, we would generate `a = b = c` as we do now).

**Comment by RealyUniqueName** [comment]
I feel like @nadako's proposal is the way to get two problems instead of one :)
I mean it's good to have cleaner setters, but that means we will have to handle this behavior fork. That could potentially affect macros. 
And since we're open to completely new properties syntax I agree with @Simn.

**Comment by nadako** [comment]
I would be fine with that, we gotta design that new property syntax ^^
